test/core/circuit.spec.js

```diff
@@ -0,0 +1,140 @@
+/* eslint max-nested-callbacks: ["error", 8] */
+/* eslint-env mocha */
+'use strict'
+
+const chai = require('chai')
+const dirtyChai = require('dirty-chai')
+const expect = chai.expect
+const parallel = require('async/parallel')
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const API = require('ipfs-api')
+const bl = require('bl')
+const PeerInfo = require('peer-info')
+const PeerId = require('peer-id')
+const multiaddr = require('multiaddr')
+const isNode = require('detect-node')
+const IPFSFactory = require('../utils/ipfs-factory-instance')
+const crypto = require('crypto')
+
+chai.use(dirtyChai)
+
+function peerInfoFromObj (obj, callback) {
+  waterfall([
+    (cb) => PeerInfo.create(PeerId.createFromB58String(obj.id), cb),
+    (peer, cb) => {
+      obj.addresses.forEach((a) => peer.multiaddrs.add(multiaddr(a)))
+      cb(null, peer)
+    }
+  ], callback)
+}
+
+describe('circuit', function () {
+  let factory
+
+  let jsRelay = new API(`/ip4/127.0.0.1/tcp/31007`)
+  let goRelay = new API(`/ip4/127.0.0.1/tcp/33027`)
+  let node1
+  let node2
+
+  let jsRelayId
+  let goRelayId
+
+  // let nodeId1
+  let nodeId2
+
+  before(function (done) {
+    // this.timeout(20 * 1000)
+    factory = new IPFSFactory()
+
+    parallel([
+      (cb) => factory.spawnNode(null, {
+        EXPERIMENTAL: {
+          Relay: {
+            Enabled: true
+          }
+        },
+        Addresses:
+          {
+            Swarm: [
+              (isNode ? `/ip4/127.0.0.1/tcp/0` : '')
+            ]
+          }
+      }, cb),
+      (cb) => factory.spawnNode(null, {
+        EXPERIMENTAL: {
+          Relay: {
+            Enabled: true
+          }
+        },
+        Addresses:
+          {
+            Swarm: [
+              (isNode ? `/ip4/127.0.0.1/tcp/0/ws` : '')
+            ]
+          }
+      }, cb)
+    ], (err, res1) => {
+      expect(err).to.not.exist()
+      node1 = res1[0]
+      node2 = res1[1]
+      parallel([
+        (cb) => jsRelay.id(cb),
+        (cb) => goRelay.id(cb),
+        (cb) => node1.id(cb),
+        (cb) => node2.id(cb)
+      ], (err, res2) => {
+        expect(err).to.not.exist()
+        parallel([
+          (cb) => peerInfoFromObj(res2[0], cb),
+          (cb) => peerInfoFromObj(res2[1], cb),
+          (cb) => peerInfoFromObj(res2[2], cb),
+          (cb) => peerInfoFromObj(res2[3], cb)
```

Comment:
Why do you need a peerInfo object? node.id() returns an array of multiaddrs with the IPFS peerId, those should be enough for a dial.

Reply:
That's because the peer object returned from the API doesn't match the `PeerInfo` structure exactly. This is what I get when calling `peer.id`

```
  {
    "id": "QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
    "publicKey": "CAASpgIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCS+u+s1cQ0HYHkVLzf8MIS4TaZq9TEl0mezvadcn2qGiOGcOMBAZ+jtH3Mcy9Bf2Y/Wozd/8lJaHG7Exnwlfxmb82vSEW2IAc7UoULG7MkUQdSC5pq7plq/RkMhzMHmiXjsnqkPdpSIi7DfNYgiNYIpd4f455pvMgcjvlpZI5BHfZtssydSkQmaUxUQughdJCqFa8RhVb4pZUEGkK7y6a286pi8NH3DXPsc3FLd6TTJOHa0T7DxErpe9+XoVN5k+mY7KbPMnnc3KhKKfEHAiRwCssqI3BTlk60k0Qet7FnYexdcdQ3d8tBkJwU9DpW/ugCUyLpytrsmEI5Yml40E2ZAgMBAAE=",
    "addresses": [
      "/ip4/127.0.0.1/tcp/10008/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/ip4/127.0.0.1/tcp/20008/ws/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ip4/127.0.0.1/tcp/10008/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ip4/127.0.0.1/tcp/20008/ws/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC",
      "/p2p-circuit/ipfs/QmRy1c5hGRVdW2PTrdNgs6SQiVVoUFfE5SmvY97EwMSHrC"
    ],
    "agentVersion": "js-ipfs",
    "protocolVersion": "9000"
  }
```

I believe that the `addresses` prop is now expected to be called `multiaddrs`, that makes it fail with `Error: peer type not recognized]` in libp2p in the `_getPeerInfo` helper. I think this is something that needs to be fixed in the api itself, but I wasn't aware of it being a problem until looking into it now.